fe/fe-core/src/main/java/com/starrocks/qe/CoordinatorPreprocessor.java

```diff
@@ -1442,6 +1452,64 @@ public static TWorkGroup prepareResourceGroup(ConnectContext connect, ResourceGr
         return resourceGroup;
     }
 
+    private List<Pair<Long, TNetworkAddress>> adaptiveChooseNodes(PlanFragment fragment, Map<Long, ComputeNode> candidates,
+                                                                  Set<TNetworkAddress> childUsedHosts) {
+        List<Pair<Long, TNetworkAddress>> childHosts = Lists.newArrayList();
+        for (Map.Entry<Long, ComputeNode> entry : candidates.entrySet()) {
+            TNetworkAddress address = entry.getValue().getAddress();
+            if (childUsedHosts.contains(address)) {
+                childHosts.add(Pair.create(entry.getKey(), address));
+            }
+        }
+
+        long maxOutputOfRightChild = fragment.getChildren().stream()
+                .sorted(Comparator.comparing(e -> e.getPlanRoot().getId().asInt())).skip(1)
+                .map(e -> e.getPlanRoot().getCardinality()).reduce(Long::max)
+                .orElse(fragment.getChild(0).getPlanRoot().getCardinality());
+        long outputOfMostLeftChild = fragment.getChild(0).getPlanRoot().getCardinality();
```

Comment:
is aways scan?

Reply:
Now, the adaptive choose instances is only used for fragment with most left plannode is not a scan . In main branch, we have refactor this scene to `RemoteFragmentAssignmentStrategy` which is more clear.